web/src/app/selection/DynamicActionField.tsx

```diff
@@ -0,0 +1,218 @@
+import { Grid, TextField } from '@mui/material'
+import DestinationField from './DestinationField'
+import { renderMenuItem } from './DisableableMenuItem'
+import AppLink from '../util/AppLink'
+
+export type StaticParams = Readonly<Record<string, string>>
+export type DynamicParams = Readonly<Record<string, ExprStringExpression>>
+
+export type Value = {
+  destType: DestinationType
+  staticParams: StaticParams
+  dynamicParams: DynamicParams
+}
+
+export function valueToActionInput(value: Value): ActionInput {
+  return {
+    dest: {
+      type: value.destType,
+      values: Object.entries(value.staticParams).map(([fieldID, value]) => ({
+        fieldID,
+        value,
+      })),
+    },
+    params: Object.entries(value.dynamicParams).map(([paramID, expr]) => ({
+      paramID,
+      expr,
+    })),
+  }
+}
+
+export function actionInputToValue(action: ActionInput): Value {
+  return {
+    destType: action.dest.type,
+    staticParams: Object.fromEntries(
+      action.dest.values.map((v) => [v.fieldID, v.value]),
+    ),
+    dynamicParams: Object.fromEntries(
+      action.params.map((p) => [p.paramID, p.expr]),
+    ),
+  }
+}
+
+export function staticToDestField(
+  staticParams?: StaticParams,
+): FieldValueInput[] {
+  if (!staticParams) return []
+  return Object.entries(staticParams).map(([fieldID, value]) => ({
+    fieldID,
+    value,
+  }))
+}
+
+export function destFieldToStatic(destFields: FieldValueInput[]): StaticParams {
+  return Object.fromEntries(destFields.map((f) => [f.fieldID, f.value]))
+}
+
+interface BaseError {
+  message: string
+}
+
+interface InputError extends BaseError {
+  section: 'input'
+  inputID: 'dest-type'
+}
+interface StaticFieldError extends BaseError {
+  section: 'static-params'
+  fieldID: string
+}
+
+interface DynamicParamError extends BaseError {
+  section: 'dynamic-params'
+  paramID: string
+}
+
+export type FormError = InputError | StaticFieldError | DynamicParamError
+
+export type DynamicActionFieldProps = {
+  value: Value | null
+  onChange: (value: Value) => void
+
+  disabled?: boolean
+
+  errors?: Array<FormError>
+}
+
+export function defaults(destTypeInfo: DestinationTypeInfo): Value {
+  const staticParams = Object.fromEntries(
+    destTypeInfo.requiredFields.map((f) => [f.fieldID, '']),
+  )
+
+  const dynamicParams = Object.fromEntries(
+    destTypeInfo.dynamicParams.map((p) => [p.paramID, `req.body.${p.paramID}`]),
+  )
+
+  return {
+    destType: destTypeInfo.type,
+    staticParams,
+    dynamicParams,
+  }
+}
+
+function isStatic(err: BaseError): err is StaticFieldError {
+  return (err as StaticFieldError).section === 'static-params'
+}
+function isInput(err: BaseError): err is InputError {
+  return (err as InputError).section === 'input'
+}
+function isDynamic(err: BaseError): err is DynamicParamError {
+  return (err as DynamicParamError).section === 'dynamic-params'
+}
+
+export default function DynamicActionField(
```

Comment:
Per our Slack convo- let's name this `DynamicActionFields`, since it's rendering more than 1 field.

```suggestion
export default function DynamicActionFields(
```

Reply:
This component operates as a single value that happens to have 1 or more fields, but not all the same, so how about `DynamicActionForm`? I think that's more in-line with how we do "manage multiple inputs" elsewhere